internal/js_library/js_library.bzl

```diff
@@ -49,96 +53,117 @@ def write_amd_names_shim(actions, amd_names_shim, targets):
                 amd_names_shim_content += "define(\"%s\", function() { return %s });\n" % n
     actions.write(amd_names_shim, amd_names_shim_content)
 
-def _impl(ctx):
-    files = []
-    typings = []
-    js_files = []
-
+def _js_library_impl(ctx):
+    direct_sources = depset(ctx.files.srcs)
+    direct_named_module_sources = depset(ctx.files.named_module_srcs)
+    sources_depsets = [direct_sources]
+    named_module_sources_depsets = [direct_named_module_sources]
+
+    # We cannot always expose the NpmPackageInfo as the linker
+    # only allow us to reference node modules from a single workspace at a time.
+    # Here we are automatically decide if we should or not including that provider
+    # by running through the sources and check if we have a src coming from an external
+    # workspace which indicates we should include the provider.
+    include_npm_package_info = False
     for src in ctx.files.srcs:
-        if src.is_source and not src.path.startswith("external/"):
-            dst = ctx.actions.declare_file(src.basename, sibling = src)
-            if ctx.attr.is_windows:
-                copy_cmd(ctx, src, dst)
-            else:
-                copy_bash(ctx, src, dst)
-            if dst.basename.endswith(".d.ts"):
-                typings.append(dst)
-            else:
-                files.append(dst)
-        elif src.basename.endswith(".d.ts"):
-            typings.append(src)
-        else:
-            files.append(src)
-
-    for p in files:
-        if p.basename.endswith(".js") or p.basename.endswith(".js.map") or p.basename.endswith(".json"):
-            js_files.append(p)
-
-    files_depset = depset(files)
+        if src.is_source and src.path.startswith("external/"):
+            include_npm_package_info = True
+            break
+
+    declarations = depset([
+        f
+        for f in ctx.files.srcs
+        if (
+               f.path.endswith(".d.ts") or
+               # package.json may be required to resolve "typings" key
+               f.path.endswith("/package.json")
```

Comment:
Would a tsconfig.json also be necessary to resolve typings? I know under `compilerOptions` you can include an array of types to include

```"compilerOptions": {
    "types": ["node", "cypress", "cypress-plugin-snapshots"]
  }```

Reply:
I believe `tsconfig.json` is only need during the `tsc` compilations. If `js_library` represents a final state of a library, the only thing needed to map types when sharing it is the `package.json` 😸